jobmaster/dm/dm_jobmaster.go

```diff
@@ -0,0 +1,246 @@
+package dm
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"go.uber.org/zap"
+
+	"github.com/hanfei1991/microcosm/executor/worker"
+	"github.com/hanfei1991/microcosm/jobmaster/dm/config"
+	"github.com/hanfei1991/microcosm/jobmaster/dm/metadata"
+	"github.com/hanfei1991/microcosm/jobmaster/dm/runtime"
+	"github.com/hanfei1991/microcosm/lib"
+	libModel "github.com/hanfei1991/microcosm/lib/model"
+	"github.com/hanfei1991/microcosm/lib/registry"
+	"github.com/hanfei1991/microcosm/model"
+	dcontext "github.com/hanfei1991/microcosm/pkg/context"
+	dmpkg "github.com/hanfei1991/microcosm/pkg/dm"
+	"github.com/hanfei1991/microcosm/pkg/p2p"
+)
+
+type JobMaster struct {
+	lib.BaseJobMaster
+
+	workerID libModel.WorkerID
+	jobCfg   *config.JobCfg
+	wg       sync.WaitGroup
+	closeCh  chan struct{}
+
+	metadata              *metadata.MetaData
+	workerManager         *WorkerManager
+	taskManager           *TaskManager
+	messageAgent          *MessageAgent
+	messageHandlerManager p2p.MessageHandlerManager
+}
+
+func RegisterWorker() {
+	constructor := func(ctx *dcontext.Context, id libModel.WorkerID, masterID libModel.MasterID, config lib.WorkerConfig) lib.WorkerImpl {
+		return NewDMJobMaster(ctx, id, masterID, config)
+	}
+	factory := registry.NewSimpleWorkerFactory(constructor, &config.JobCfg{})
+	registry.GlobalWorkerRegistry().MustRegisterWorkerType(lib.DMJobMaster, factory)
+}
+
+func NewDMJobMaster(ctx *dcontext.Context, workerID libModel.WorkerID, masterID libModel.MasterID, conf lib.WorkerConfig) *JobMaster {
+	log.L().Info("new dm jobmaster", zap.String("id", workerID))
+	jm := &JobMaster{
+		workerID: workerID,
+		jobCfg:   conf.(*config.JobCfg),
+		closeCh:  make(chan struct{}),
+	}
+
+	// TODO: we should expose the message handler register Func in base master.
+	// nolint:errcheck
+	ctx.Deps().Construct(func(m p2p.MessageHandlerManager) (p2p.MessageHandlerManager, error) {
+		jm.messageHandlerManager = m
+		return m, nil
+	})
+	return jm
+}
+
+func (jm *JobMaster) createComponments() error {
+	taskStatus, workerStatus, workerHandles, err := jm.getInitStatus()
+	if err != nil {
+		return err
+	}
+	jm.metadata = metadata.NewMetaData(jm.ID(), jm.MetaKVClient())
+	jm.messageAgent = NewMessageAgent(workerHandles, jm.ID(), jm.BaseJobMaster)
+	jm.taskManager = NewTaskManager(taskStatus, jm.metadata.JobStore(), jm.messageAgent)
+	jm.workerManager = NewWorkerManager(workerStatus, jm.metadata.JobStore(), jm.messageAgent, &CheckpointAgentImpl{})
+	return nil
+}
+
+func (jm *JobMaster) InitImpl(ctx context.Context) error {
+	log.L().Info("initializing the dm jobmaster", zap.String("id", jm.workerID), zap.String("jobmaster id", jm.JobMasterID()))
+	if err := jm.createComponments(); err != nil {
+		return err
+	}
+	if err := jm.registerMessageHandler(ctx); err != nil {
+		return err
+	}
+	return jm.taskManager.OperateTask(ctx, Create, jm.jobCfg, nil)
```

Comment:
should we split Job to Task first?

Reply:
I think job is enough, because it is a submit-job command